### ISSUE TITLE
should_run_async detects simple list comprehensions as async cells and cause bugs

### ISSUE THREAD

**Issue by JoaoFelipe** [ipython/ipython#12422]
Trying to understand [this bug](https://github.com/jupyter/notebook/issues/5568) that occurs in both Jupyter Notebook and Jupyter Lab in Python >= 3.8, I noticed that it is related to the async engine of IPython. Code that should execute synchronously is often detected as async:
```
In [1]: get_ipython().should_run_async('[x for x in l]')                        
Out[1]: True
```
It occurs for list comprehensions, set comprehensions, dict comprehensions and generator expressions, but it does not occur for these elements when they are inside a function definition (even when it is called):
```
In [2]: get_ipython().should_run_async('(lambda: [x for x in l])()')                        
Out[2]: False
```

I also checked other ast elements that make this function return True and the only ones that do are the expected ones: `async for`, `async with`, and `await`.

(Fixing this function may solve the bug that I reported on the jupyter/notebook repository, but I think there is something else going on with async executions: when I was developing a Jupyter Lab extension, one of my callbacks for `kernel.requestExecute` received the async response of a notebook cell. Even in an async execution, I think the original cell should receive the response and update its result)

Other information:
```
python -c "import IPython; print(IPython.sys_info())"
{'commit_hash': 'bb015d48b',
 'commit_source': 'repository',
 'default_encoding': 'UTF-8',
 'ipython_path': '/home/joao/projects/ipython/IPython',
 'ipython_version': '8.0.0.dev',
 'os_name': 'posix',
 'platform': 'Linux-4.15.0-20-generic-x86_64-with-debian-buster-sid',
 'sys_executable': '/home/joao/anaconda3/bin/python',
 'sys_platform': 'linux',
 'sys_version': '3.7.3 (default, Mar 27 2019, 22:11:17) \n[GCC 7.3.0]'}
```

**Comment by Carreau** [comment]
Do you have extensions ? Cause ti definitively returns "False" for me. See below, and try to following to see if you have any defferences.

```
In [6]: get_ipython().should_run_async('[x for x in l]')
Out[6]: False

In [7]: get_ipython().transform_cell('[x for x in l]')
Out[7]: '[x for x in l]\n'

In [8]: from IPython.core.async_helpers import _should_be_async

In [9]: _should_be_async('[x for x in l]')
Out[9]: False
````

**Comment by JoaoFelipe** [comment]
Here are my results:
```
Python 3.8.3 (default, May 19 2020, 18:47:26) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.0.0.dev -- An enhanced Interactive Python. Type '?' for help.

In [1]: get_ipython().should_run_async('[x for x in l]')                        
Out[1]: True

In [2]: get_ipython().transform_cell('[x for x in l]')                          
Out[2]: '[x for x in l]\n'

In [3]: from IPython.core.async_helpers import _should_be_async                 

In [4]: _should_be_async('[x for x in l]')                                      
Out[4]: True
```

Note that it only occurs on Python 3.8. Python 3.7 works:
```
Python 3.7.3 (default, Mar 27 2019, 22:11:17) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: get_ipython().should_run_async('[x for x in l]')                        
Out[1]: False

In [2]: get_ipython().transform_cell('[x for x in l]')                          
Out[2]: '[x for x in l]\n'

In [3]: from IPython.core.async_helpers import _should_be_async                 

In [4]: _should_be_async('[x for x in l]')                                      
Out[4]: False
```

**Comment by JoaoFelipe** [comment]
It also occurs on 3.10 (I used the binder provided at https://github.com/gvanrossum/patma):
![image](https://user-images.githubusercontent.com/327789/86542785-2676c300-beef-11ea-84bf-e6bf0a545084.png)

**Comment by Carreau** [comment]
https://bugs.python.org/issue41218

**Comment by Carreau** [comment]
introduced by https://bugs.python.org/issue39562

**Comment by Carreau** [comment]
Closed in upstream Python, will be in Python 3.8.4.
